libnavigation-core/src/main/java/com/mapbox/navigation/core/routerefresh/RouteRefreshController.kt

```diff
@@ -53,27 +58,93 @@ internal class RouteRefreshController(
      * Cancel old timers, and cancel pending requests.
      */
     fun stop() {
-        routerRefreshTimer.stopJobs()
-        currentRequestId?.let {
-            directionsSession.cancelRouteRefreshRequest(it)
-            currentRequestId = null
-        }
+        currentRefreshRoutesJob?.cancel()
+        currentRefreshRoutesJob = null
     }
 
-    private fun refreshRoute(
+    private suspend fun refreshRoutes(
         route: NavigationRoute,
-        onRoutesRefreshed: (List<NavigationRoute>) -> Unit
-    ) {
+        onRouteRefresh: (NavigationRoute) -> Unit
+    ): Unit = coroutineScope {
+        var cleanupOutdatedTrafficJob: Job? = null
+        var refreshRouteJob: Job? = null
+        while (true) {
+            cleanupOutdatedTrafficJob = cleanupOutdatedTrafficJob ?: launch {
+                delay(trafficAnnotationsExpirationTime())
+                val routeWithoutTrafficAnnotations = removeTrafficAnnotationsOrNull(route)
+                if (routeWithoutTrafficAnnotations != null) {
+                    onRouteRefresh(routeWithoutTrafficAnnotations)
+                }
+            }
+            delay(routeRefreshOptions.intervalMillis)
+            refreshRouteJob?.cancel()
+            refreshRouteJob = launch {
+                val refreshedRoute = refreshRouteOrNull(route)
+                if (refreshedRoute != null) {
+                    cleanupOutdatedTrafficJob?.cancel()
+                    cleanupOutdatedTrafficJob = null
+                    onRouteRefresh(refreshedRoute)
+                }
+            }
+        }
+    }
+
+    private fun trafficAnnotationsExpirationTime() = routeRefreshOptions.intervalMillis * 3
+
+    private fun combineWithAlternativeRoutesFromDirectionsSession(
+        newRoute: NavigationRoute
+    ): List<NavigationRoute> {
+        val directionsSessionRoutes = directionsSession.routes.toMutableList()
+        if (directionsSessionRoutes.isNotEmpty()) {
+            directionsSessionRoutes[0] = newRoute
+        }
+        return directionsSessionRoutes
+    }
+
+    private fun removeTrafficAnnotationsOrNull(route: NavigationRoute): NavigationRoute? {
+        val currentLegIndex = currentLegIndexProvider()
+        val routeLegs = route.directionsRoute.legs() ?: return null
+        return route.updateLegAnnotations(
+            currentLegIndex,
+            routeLegs.mapIndexed { index, leg ->
+                if (index == currentLegIndex) {
+                    leg.annotation()
+                        ?.toBuilder()
+                        ?.congestion(null)
+                        ?.congestionNumeric(null)
```

Comment:
Annotations shouldn't be removed but rather replaced with a matching-size collections of `"unknown"` (and numeric equivalent) congestion segments. That's to keep consistent with `RouteOptions` that did specify congestion to be returned and it would mean a mismatch if they were suddenly gone.

Reply:
Replaced by "unkown" and null